EIPS/eip-5988.md

```diff
@@ -0,0 +1,163 @@
+Moreover it is compatible with all major proof systems (SNARKs, STARKs, Bulletproofs, etc...).
+This makes it a good candidate for a precompile that can be used by many different ZK-Rollups.
+
+An important point to note is that ZK rollups using Poseidon have chosen different sets of parameters, which makes it harder to build a single precompile for all of them.
+
+However, we can still build a generic precompile that supports arbitrary parameters, and allow the ZK rollups to choose the parameters they want to use.
+
+This is the approach that we have taken in this EIP.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+The `POSEIDON` precompile MUST be available at address `0xPOSEIDON_PRECOMPILE_ADDRESS`.
+
+The precompile MUST be activated at `FORK_BLOCK_NUMBER`.
+
+Here are the Poseidon parameters that the precompile will support:
+
+- `p`: the prime field modulus.
+- `security_level`: the security level measured in bits. Denoted `M` in the Poseidon paper.
+- `alpha`: the power of S-box.
+- `input_rate`: the size of input.
+- `t`: the size of the state.
+- `full_round`: the number of full rounds. Denoted as `R_F` in the Poseidon paper.
+- `partial_round`: the number of partial rounds. Denoted as `R_P` in the Poseidon paper.
+
+The input to the precompile MUST be encoded as follows:
+
+- `p`: 32 bytes
+- `security_level`: 2 bytes
+- `alpha`: 1 byte
+- `input_rate`: 2 bytes
+- `t`: 1 byte
+- `full_round`: 1 byte
+- `partial_round`: 1 byte
+- `input`: `input_rate` \* 32 bytes
+
+```
+[32 bytes for p][2 bytes for security_level][1 byte for alpha][2 bytes for input_rate][1 byte for t][1 byte for full_round][1 byte for partial_round][input_rate * 32 bytes for input]
+```
+
+The precompile should compute the hash function as [specified in the Poseidon paper](https://eprint.iacr.org/2019/458.pdf) and return hash output.
+
+### Example Usage in Solidity
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `poseidon_hash` function.
+
+```solidity
+// TODO: Add solidity example
+```
+
+### Gas Costs
+
+```
+TODO: Fill gas costs section and add benchmarks
+```
+
+## Rationale
+
+TODO: Add rationale
+
+## Backwards Compatibility
+
+There is very little risk of breaking backwards-compatibility with this EIP, the sole issue being if someone were to build a contract relying on the address at `0xPOSEIDON_PRECOMPILE_ADDRESS` being empty. The likelihood of this is low, and should specific instances arise, the address could be chosen to be any arbitrary value with negligible risk of collision.
+
+## Test Cases
+
+The Poseidon reference implementation contains test vectors that can be used to test the precompile.
+Those tests are available [here](https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/test_vectors.txt).
```

Comment:
Tests and reference implementation should be located either within the EIP text or in the `assets/EIP-5988` directory.